syntax = "proto3";

package template;

//import "google/protobuf/empty.proto"; 
import "google/protobuf/any.proto";
 
service Template{
 rpc Get(GetRequest) returns (GetResponse) {}
 rpc List(ListRequest) returns (ListResponse) {}
 rpc Delete(DeleteRequest) returns (DeleteResponse) {}
 rpc Update(UpdateRequest) returns (UpdateResponse) {}
 rpc CreateFromDesktop(CreateFromDesktopRequest) returns (CreateFromDesktopResponse) {}
 
 rpc BaseGet(BaseGetRequest) returns (BaseGetResponse) {}
 rpc BaseList(BaseListRequest) returns (BaseListResponse) {}
}

message HardwareNew{
 int32 vcpus=1;
 int32 memory=2;
 string boot_disk_rpath=3;
 string boot_disk_bus=4;
 int32  boot_disk_size=5; 
 repeated string videos=6;
 repeated string graphics=7;
 repeated string boots=8;
 repeated string interfaces=9;
 repeated string isos=10;
 repeated string floppies=11;
}

message HardwareDerived{
 int32 vcpus=1;
 int32 memory=2;
 string boot_disk_rpath=3;
 string boot_disk_bus=4; 
 repeated string videos=5;
 repeated string graphics=6;
 repeated string boots=7;
 repeated string interfaces=8;
 repeated string isos=9;
 repeated string floppies=10;
}

message HardwareUpdate{
 int32 vcpus=1;
 int32 memory=2;
 string boot_disk_bus=7; 
 repeated string videos=3;
 repeated string graphics=4;
 repeated string boots=5;
 repeated string interfaces=6;
 repeated string isos=8;
 repeated string floppies=9;
}
    
message Viewer{
 string hostname=1;
 string hostname_external=2;
 int32 port_spice=5;
 int32 port_spice_ssl=6;
 int32 port_vnc=7;
 int32 port_vnc_websocket=8;
 string passwd=9;
 string client_addr=10;
 float client_since=11;
}
    

    // TEMPLATE //

message GetRequest{
 string template_id=1;
}

message GetResponse{
  repeated google.protobuf.Any template = 1;
}

message ListRequest{}

message ListResponse{
 repeated string templates = 1;
}

message DeleteRequest{
 string template_id = 1;
}

message DeleteResponse{
    enum State{
     DELETED = 0;
     FAILED = 1;
     UNKNOWN = 2;
    }
 State state = 1;
 string detail = 2;
}

message UpdateRequest{
 string template_id=1;
 HardwareUpdate hardware=2;
}

message UpdateResponse{
    enum State{
     STOPPED= 0;
     FAILED = 1;
     UNKNOWN = 2;
    }
 State state = 1;
 string detail = 2;
}

message CreateFromDesktopRequest{
 string desktop_id=1;
 string template_id=2;
 HardwareDerived hardware=3;
 repeated string next_actions = 4;
}

message CreateFromDesktopResponse{
    enum State{
     STOPPED= 0;
     FAILED = 1;
     UNKNOWN = 2;
    }
 State state = 1;
 string detail = 2;
}


    // BASE //

message BaseGetRequest{
 string base_id=1;
}

message BaseGetResponse{
  repeated google.protobuf.Any base = 1;
}

message BaseListRequest{}

message BaseListResponse{
 repeated string bases=1;
}

